bin/picom-trans

```diff
@@ -46,194 +55,325 @@ print_usage() {
 
-# Variables
-active=
-wprefix=
-window=
-opacity=
-cur=
-action=
-treeout=
-wid=
-topmost=
-lineno=
-option=
-v=
-
-# We make getopts stop on any argument it doesn't recognize
-# or errors on. This allows for things like `picom-trans -5`
-# as well as `picom-trans -c +5 -s` (contrived example).
-while test $# -gt 0; do
-  # Reset option index
-  OPTIND=1
+      # Turn into short form
+      case "$1" in
+        (--help|--opacity|--get|--delete|--toggle|--reset|--select|--current|--name|--window)
+          ARG=${1#-}                          # remove one '-' from prefix
+          ARG="$(echo "$ARG" | cut -c -2)"    # get first two characters
+          shift && set -- "$ARG" "$@"
+      esac
 
-  # Read options
-  while getopts ':hscrtdgn:w:o:-:' option "$@"; do
-    if test "$option" = '-'; then
-      case "$OPTARG" in
-        help | select | current | reset | toggle | delete | get)
-          v=''
-        ;;
-        name | window | opacity)
-          eval v=\$$OPTIND
-          OPTIND=$((OPTIND + 1))
-        ;;
-        name=* | window=* | opacity=*)
-          v=$(echo "$OPTARG" | sed -E 's/^[^=]+=//')
-        ;;
-        *)
-          echo "$0: illegal option $OPTARG" >& 2
+      # If the argument still starts with --, it is an invalid argument
+      case "$1" in
+        (--*)
+          echo "$EXE_NAME: illegal option $1" >&2
           exit 1
-        ;;
       esac
-      option=$(echo "$OPTARG" | cut -c 1)
-      OPTARG=$v
     fi
-    case "$option" in
-      h) print_usage; exit 0 ;;
-      s) wprefix=''; window='' ;;
-      c)
-        active=$(xprop -root -notype _NET_ACTIVE_WINDOW \
-          | grep -Eo '0x[[:xdigit:]]+' | head -n 1)
-        wprefix='-id'; window=$active
-      ;;
-      r) action='reset' ;;
-      t) action='toggle' ;;
-      d) action='delete' ;;
-      g) action='get' ;;
-      n) wprefix='-name'; window=$OPTARG ;;
-      w) wprefix='-id'; window=$OPTARG ;;
-      o) opacity=$OPTARG ;;
-      \?) break ;;
-    esac
-  done
+    #### [END] Convert GNU longopts to POSIX equivalents ####
 
-  # Read positional arguments
-  shift $((OPTIND - 1))
-  test -n "$1" && opacity=$1 && shift
-done
 
-# clean up opacity. xargs == a poor man's trim.
-opacity=$(echo "$opacity" | xargs | sed 's/%//g')
+    #### [START] Prepend '-o' to  standalone opacity values ####
+    # Iterate over every argument and check if it is an opacity without the -o
+    # option in the previous argument. If so, then prepend the -o option.
+    # e.g. Turn this -
+    #     picom-trans -c +10 -s
+    # into this -
+    #     picom-trans -c -o +10 -s
+    #
+    # NOTE: Don't touch arguments that are preceded by -n or -w
+    #       i.e. don't touch arguments that take a value.
+    # i.e. This -
+    #     picom-trans -w 75 -o 90
+    # should NOT be turned into this -
+    #     picom-trans -w -o 75 -o 90
+    if echo "$1" | grep -qE '^[+-]?[[:digit:]]+%?$' && \
+      ! eval echo "\"\${$#}\"" | grep -q '^-[hdtrgsc]*[own]$'
```

Comment:
`echo` does not work well, if the argument we want to check is `-n`: https://github.com/yshui/picom/pull/634#discussion_r661498300

Reply:
Does using the following diff resolve the problem? -
```diff
diff --git a/bin/picom-trans b/bin/picom-trans
index 9f1e44f..f034910 100755
--- a/bin/picom-trans
+++ b/bin/picom-trans
@@ -118,8 +118,8 @@ parse_args()
     # should NOT be turned into this -
     #     picom-trans -w -o 75 -o 90
     if echo "$1" | grep -qE '^[+-]?[[:digit:]]+%?$' && \
-      ! eval echo "\"\${$#}\"" | grep -q '^-[hdtrgsc]*[own]$'
-    # NOTE: eval echo \${$#} means 'echo the last argument'
+      ! eval printf %s \$$# | grep -q '^-[hdtrgsc]*[own]$'
+    # NOTE: eval printf %s ${$#} means 'echo the last argument'
     # NOTE: The letters inside the first  square brackets (ie. hdtrgsc) are
     #       the same as those in the getopts argument, minus those that are
     #       followed by a ':'
```

Reply:
While the above diff should work, I prefer using explicit quotation marks as in the diff posted before:
```diff
diff --git a/bin/picom-trans b/bin/picom-trans
index a6868e1..b32fddf 100755
--- a/bin/picom-trans
+++ b/bin/picom-trans
@@ -117,9 +117,9 @@ parse_args()
     #     picom-trans -w 75 -o 90
     # should NOT be turned into this -
     #     picom-trans -w -o 75 -o 90
-    if echo "$1" | grep -qE '^[+-]?[[:digit:]]+%?$' && \
-      ! eval echo "\"\${$#}\"" | grep -q '^-[hdtrgsc]*[own]$'
-    # NOTE: eval echo \${$#} means 'echo the last argument'
+    if printf '%s' "$1" | grep -qE '^[+-]?[[:digit:]]+%?$' && \
+      ! eval "printf '%s' \"\${$#}\"" | grep -q '^-[hdtrgsc]*[own]$'
+    # NOTE: eval printf '%s' "\${$#}" means 'echo the last argument'
     # NOTE: The letters inside the first  square brackets (ie. hdtrgsc) are
     #       the same as those in the getopts argument, minus those that are
     #       followed by a ':'
```